### ISSUE TITLE
CS8625 when assigning null to MaybeNullWhen(false) out parameter then returning false

### ISSUE THREAD

**Issue by drewnoakes** [dotnet/roslyn#36917]
**Version Used**: 3.3.0-beta1-19327-04

**Steps to Reproduce**:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.2</TargetFramework>
    <LangVersion>preview</LangVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Net.Compilers.ToolSet" Version="3.3.0-beta1-19328-01" />
  </ItemGroup>
</Project>
```

```c#
#nullable enable

using System.Diagnostics.CodeAnalysis;

namespace NullableAnnotationSpike
{
    class C
    {
        string? M()
        {
            if (TryGet1(1, out var s1))
                return s1.ToUpper();

            if (TryGet2(1, out var s2))
                return s2.ToUpper();

            return s1.ToUpper()  // Expected CS8602
                 + s2.ToUpper(); // Expected CS8602
        }

        bool TryGet1(int _, [MaybeNullWhen(false)] out string s)
        {
            s = null;           // Unexpected CS8625
            return false;
        }

        bool TryGet2(int _, [NotNullWhen(true)] out string? s)
        {
            s = null;
            return false;
        }
    }
}

namespace System.Diagnostics.CodeAnalysis
{
    [AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
    internal sealed class MaybeNullWhenAttribute : Attribute
    {
        public MaybeNullWhenAttribute(bool returnValue) => ReturnValue = returnValue;

        public bool ReturnValue { get; }
    }

    [AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
    public sealed class NotNullWhenAttribute : Attribute
    {
        public NotNullWhenAttribute(bool returnValue) => ReturnValue = returnValue;

        public bool ReturnValue { get; }
    }
}
``` 

**Expected Behavior**:

No diagnostic assigning `s = null` in `TryGet1` before returning `false`.

**Actual Behavior**:

Diagnostic.

**Comment by agocke** [comment]
@drewnoakes This looks wrong:

```C#
        bool TryGet1(int _, [MaybeNullWhen(false)] out string s)
        {
            s = null;           // Unexpected CS8625
            return false;
        }
```

If your type is a known reference type and can contain null, it should be marked nullable. Your second version looks correct.

However, I think for the broader question, the attributes currently don't affect the method bodies. @jcouv Is that still correct?

**Comment by drewnoakes** [comment]
@agocke in that case I can't see when you'd use `MaybeNullWhen`.

If I make the `out` parameter's type generic then I see CS8653 (A default expression introduces a null value when 'T' is a non-nullable reference type).

```c#
        bool TryGet3<T>(int _, [MaybeNullWhen(false)] out T t)
        {
            t = default;        // CS8653
            return false;
        }

        bool TryGet4<T>(int _, [NotNullWhen(true)] out T t)
        {
            t = default;        // CS8653
            return false;
        }
```

**Comment by agocke** [comment]
Unconstrained generic is different -- in that case you don't know if the type is a reference type. That's when I would expect MaybeNull to be used.

In this case, however, I think the "output" attributes still wouldn't affect the method body. They are primarily for consumers of the API.

**Comment by drewnoakes** [comment]
I'm not clear on how a developer would avoid the CS8653 warnings on assignment in `TryGet3` and `TryGet4`. It's not possible to declare the type as `T?` without a `class` constraint. Adding `[MaybeNull]` doesn't help.

**Comment by agocke** [comment]
Right, in this case you will have to suppress the warning. We don't currently use the attributes to affect method bodies, but even if we implement that behavior, I think tracking `[MaybeNullIf` would be difficult.

**Comment by jcouv** [comment]
> the attributes currently don't affect the method bodies. @jcouv Is that still correct?

Yes, that is correct at this point and likely for C# 8 RTM. This is tracked by https://github.com/dotnet/roslyn/issues/36039

**Comment by jcouv** [comment]
This was fixed in 16.5p3 in PR https://github.com/dotnet/roslyn/pull/41098 (Enforce conditional attributes within method bodies)
I'll go ahead and close. Thanks
